modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/criteria_builder/ContributorsBuilder.es.js

```diff
@@ -0,0 +1,282 @@
+					...prevState,
+					contributors
+				};
+			}
+		);
+	}
+
+	render() {
+		const {
+			propertyGroups,
+			supportedConjunctions,
+			supportedOperators,
+			supportedPropertyTypes
+		} = this.props;
+		const currentEditing = this.state.editing;
+		const selectedContributor = this.state.contributors[currentEditing];
+		const selectedProperty = selectedContributor && propertyGroups.find(c => selectedContributor.propertyKey === c.propertyKey);
+
+		return (
+			<div className={this.classes}>
+				<div className="criteria-builder-section-main">
+					{
+						this.state.contributors.map(
+							(criteria, i) => {
+								return (
+									<React.Fragment key={i}>
+										<div className={'sheet-lg'}>
+											{
+												(i !== 0) &&
+												<React.Fragment>
+													<Conjunction
+														className={'ml-0'}
+														conjunctionName={criteria.conjunctionId}
+														editing={true}
+														supportedConjunctions={supportedConjunctions}
+													/>
+
+													<input
+														id={criteria.conjunctionInputId}
+														readOnly
+														type="hidden"
+														value={criteria.conjunctionId}
+													/>
+												</React.Fragment>
+											}
+										</div>
+
+										<CriteriaBuilder
+											criteria={criteria.criteriaMap}
+											editing={currentEditing === i}
+											id={i}
+											initialQuery={criteria.query}
+											inputId={criteria.inputId}
+											modelLabel={criteria.modelLabel}
+											onChange={this.onCriteriaChange}
+											onEditToggle={this.onCriteriaEdit}
+											propertyKey={criteria.propertyKey}
+											supportedConjunctions={supportedConjunctions}
+											supportedOperators={supportedOperators}
+											supportedProperties={criteria.properties}
+											supportedPropertyGroups={this.props.propertyGroups.map(
+												p => ({
+													label: p.name,
+													value: p.propertyKey
+												})
+											)}
+											supportedPropertyTypes={supportedPropertyTypes}
+										/>
+
+										<div className="form-group">
+											<input
+												className="field form-control"
+												data-testid={criteria.inputId}
+												id={criteria.inputId}
+												name={criteria.inputId}
+												readOnly
+												type="hidden"
+												value={criteria.query}
+											/>
+										</div>
+									</React.Fragment>
+								);
+							}
+						)
+					}
+				</div>
+
+				<div className="criteria-builder-section-sidebar">
+					{
+						<CriteriaSidebar
+							propertyKey={selectedProperty && selectedProperty.propertyKey}
+							supportedProperties={selectedProperty && selectedProperty.properties}
+							title={
+								sub(
+									Liferay.Language.get('x-properties'),
+									[selectedProperty && selectedProperty.name]
+								)
+							}
+						/>
+					}
+				</div>
+			</div>
+		);
+	}
+}
+
+const conjunctions = PropTypes.shape(
```

Comment:
For all of these shapes, can you make the variable name something like `conjunctionsShape`, `initialContributorShape`, etc. This way we will avoid any accidental collisions in variable names and its always better to be more explicit than less.